consensus/src/messages/handlers.rs

```diff
@@ -476,3 +476,28 @@ impl Handle<ResponseTransactionReceiptsByAddress, Arc<RwLock<Blockchain>>>
         ResponseTransactionReceiptsByAddress { receipts }
     }
 }
+
+#[cfg(feature = "full")]
+impl Handle<ResponseAccountsProof, Arc<RwLock<Blockchain>>> for RequestAccountsProof {
+    fn handle(&self, blockchain: &Arc<RwLock<Blockchain>>) -> ResponseAccountsProof {
+        let blockchain = blockchain.read();
+
+        // We only prove accounts that exist in our current state
+        let proof = blockchain.get_accounts_proof(&self.addresses);
+
+        if proof.is_none() {
+            // If we could not generate a proof we response an empty result
+            return ResponseAccountsProof {
+                proof: None,
+                block_hash: None,
+            };
+        }
+
+        let block_hash = blockchain.head_hash();
```

Comment:
Is this guaranteed to be the same hash that is used to generate the proof? No chance that the head is updated in this blockchain-read in the meantime?

Reply:
Yup, the read lock is taken while this happens